components/layout_2020/flow/mod.rs

```diff
@@ -143,21 +145,91 @@ impl BlockContainer {
         writing_mode: WritingMode,
     ) -> ContentSizes {
         match &self {
-            Self::BlockLevelBoxes(boxes) if layout_context.use_rayon => boxes
-                .par_iter()
-                .map(|box_| {
-                    box_.borrow_mut()
-                        .inline_content_sizes(layout_context, writing_mode)
-                })
-                .reduce(ContentSizes::zero, ContentSizes::max),
-            Self::BlockLevelBoxes(boxes) => boxes
-                .iter()
-                .map(|box_| {
-                    box_.borrow_mut()
-                        .inline_content_sizes(layout_context, writing_mode)
-                })
-                .reduce(ContentSizes::max)
-                .unwrap_or_else(ContentSizes::zero),
+            Self::BlockLevelBoxes(boxes) => {
+                let get_box_info = |box_: &ArcRefCell<BlockLevelBox>| {
+                    let size = box_
+                        .borrow_mut()
+                        .inline_content_sizes(layout_context, writing_mode);
+                    if let BlockLevelBox::OutOfFlowFloatBox(ref float_box) = *box_.borrow_mut() {
+                        let style_box = match &float_box.contents {
+                            IndependentFormattingContext::Replaced(ref replaced) => {
+                                replaced.style.get_box()
+                            },
+                            IndependentFormattingContext::NonReplaced(ref non_replaced) => {
+                                non_replaced.style.get_box()
+                            },
+                        };
+                        (size, style_box.float, style_box.clear)
+                    } else {
+                        // The element may in fact have clearance, but the logic below ignores it,
+                        // so don't bother retrieving it from the style.
+                        (size, Float::None, Clear::None)
+                    }
+                };
+                struct AccumulatedData {
+                    /// The maximum size seen so far, not including trailing floats.
+                    max_size: ContentSizes,
+                    /// The size of the trailing floats with 'float: left'.
+                    left_floats: ContentSizes,
+                    /// The size of the trailing floats with 'float: right'.
+                    right_floats: ContentSizes,
+                }
+                impl AccumulatedData {
+                    fn max_size_including_floats(&self) -> ContentSizes {
```

Comment:
Maybe this should be called something like `max_size_including_uncleared_floats`?